CFE-3846: Now you can specify version when adding modules

larsewi opened the pull request.

larsewi committed:
Add cfbs.json to .gitignore

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Add build output to .gitignore

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

olehermanse commented on cfbs/core.py:
```diff
@@ -123,36 +127,51 @@ def data(self) -> dict:
+            if version != "latest":
+                # Change fields according to versions.json
+                versions = get_json(_VERSION_INDEX)
+                object.update({"version": version})
```
```suggestion
                object["version"] = version
```

olehermanse commented on cfbs/core.py:
```diff
@@ -123,36 +127,51 @@ def data(self) -> dict:
+            if version != "latest":
+                # Change fields according to versions.json
+                versions = get_json(_VERSION_INDEX)
+                object.update({"version": version})
+                for key, val in versions[name][version].items():
+                    if key in object:
+                        object.update({key: val})
```
```suggestion
                new_version = versions[name][version]
                object.update({"version": version})
                desired = ["commit", "version", "steps", "subdirectory", "by", "repo", "dependencies"]
                new_values = {k,v in new_version.items() if k in desired}
                object.update(new_values)
```

larsewi committed:
Moved Index class to index.py

The file core.py is starting to become unmaintainable due to its size,
therefore we moved the Index class to it's on file.

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Moved CFBSJson class to cfbs_json.py

Moved CFBSJson class from core.py to cfbs_json.py in order to improve
maintainability.

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Change name of core.py to cfbs_config.py

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Remove unused imports

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Move constant _VERSION_INDEX to index.py

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Changed name of parameter in CFBSConfig._add_modules

Changed name of parameter `names` in `CFBSConfig._add_modules` to
`to_add`.

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Add modules with specific versions

You now can add modules with a specific version using the following
format: `cfbs add NAME[@VERSION]`

Ticket: CFE-3846
Changelog: Body
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Test for adding module of specific version

Ticket: CFE-3846
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Fix 'distutils.spawn' error while running pytests

For some reason when running pytest in workflow, it errors with the
message "module 'distutils' has no attribute 'spawn'". This apparently
fixes the issue. Don't ask me why.

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

vpodzime reviewed (commented):
Looks good to me otherwise.

larsewi commented on cfbs/cfbs_config.py:
```diff
@@ -145,27 +147,26 @@ def _convert_added_by(added_by, to_add):
-                new_added_by = added_by[module["name"]]
-                if new_added_by == "cfbs add":
-                    module["added_by"] = "cfbs add"
+        for req, mod in itertools.product(requested, self["build"]):
```
You're right, I'll fix!

larsewi commented on cfbs/cfbs_config.py:
```diff
@@ -208,30 +208,30 @@ def _add_without_dependencies(self, modules):
     def _add_modules(
         self,
-        names: list,
+        to_add: list,
```
Good catch 👍

larsewi commented on cfbs/module.py:
```diff
@@ -0,0 +1,53 @@
+        super().update(m)
+
+    @staticmethod
+    def allowed_keys():
```
> Can we use a traditional class with attributes?

Traditional class is a bit wonky in that some attributes are optional while others are mutually exclusive.

> Can we solve this with properties?

Yes, but it requires a lot of code for each attribute. Therefore I used the `__getattr__` and `__setattr__` functions instead to implement functionality for all attributes at once.

olehermanse commented on cfbs/cfbs_config.py:
```diff
@@ -1,6 +1,7 @@
 #!/usr/bin/env python3
 import os
 import logging as log
+import itertools
```
Unused import

larsewi replied:
Fixed

olehermanse commented on test/shell/014_add_nonexistent.sh:
```diff
@@ -7,4 +7,4 @@ touch cfbs.json && rm cfbs.json
 cfbs init
 ! ( cfbs add bollocks > output.log 2>&1 )
-grep -F "Error: Module 'bollocks' does not exist" output.log
+grep -F "Error: Module 'bollocks' version 'latest' does not exist" output.log
```
IMO, this error message was nicer before and we could make it only print version in case it's not latest.

larsewi replied:
Agreed 👍

olehermanse commented on cfbs/index.py:
```diff
@@ -0,0 +1,161 @@
+        else:
+            name = module
+            version = None
+        if os.path.exists(name):
+            return True
+        else:
+            if not version:
+                return name in self
+            else:
+                versions = get_json(_VERSION_INDEX)
+                return name in versions and version in versions[name]
```
Early return makes these else's unnecessary:

```suggestion
        if os.path.exists(name):
            return True
        if not version:
            return name in self
        versions = get_json(_VERSION_INDEX)
        return name in versions and version in versions[name]
```

https://refactoring.guru/replace-nested-conditional-with-guard-clauses

olehermanse reviewed (approved):
Looks good, thanks for the improvements and adding tests :)

I left some comments but feel free to address them in a follow-up.

vpodzime commented on cfbs/cfbs_config.py:
```diff
@@ -145,27 +147,27 @@ def _convert_added_by(added_by, to_add):
-                if new_added_by == "cfbs add":
-                    module["added_by"] = "cfbs add"
+        for req in requested:
+            for mod in self["build"]:
```
Would it be possible to do `if req["name"] in self["build"]:`?

olehermanse replied:
`self["build"]` is a list (JSON array) so this doesn't work, but we could add a helper, `self.already_added(name)`.

vpodzime commented on cfbs/module.py:
```diff
@@ -0,0 +1,74 @@
+from cfbs.pretty import pretty
+
+
+class Module(OrderedDict):
```
Please double-check the methods in the picture are OK for `Module` objects:
![image](https://user-images.githubusercontent.com/1044278/158221326-7f26ec51-4ca4-4ffb-a3af-607dabc72379.png)

vpodzime replied:
Also, what about `del module["name"]`?

larsewi replied:
N/A

vpodzime reviewed (commented):
Looks good to me otherwise.

larsewi committed:
Added unit tests for Module class

Ticket: CFE-3846
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

vpodzime commented on cfbs/module.py:
```diff
@@ -0,0 +1,68 @@
+    def to_dict(self):
+        keys = Module.attributes()
+        vals = [getattr(self, attr) for attr in keys]
+        return OrderedDict({key: val for (key, val) in zip(keys, vals) if val})
```
You are creating an OrderedDict from a normal dict here. That doesn't make much sense. I think you need to use the tuples directly.

vpodzime commented on cfbs/index.py:
```diff
@@ -97,35 +98,65 @@ def data(self) -> dict:
+                versions = get_json(_VERSION_INDEX)
+                new_values = versions[name][version]
+                specifics = {
+                    k: v for (k, v) in new_values.items() if k in Module.keys()
```
This should be `Module.attributes()`, right?

larsewi replied:
Good catch 🙌

vpodzime reviewed (commented):
Looks good to me. Although it's a bit of a mystery to me how the tests can pass with the `Module.keys()` bug.

vpodzime commented:
> Although it's a bit of a mystery to me how the tests can pass with the Module.keys() bug.

Oh, mystery solved -- the don't pass. Only on the commit adding the functionality because the test is in the commit following it.

vpodzime reviewed (approved):
Let's not waste any extra time on this.
